linux/mcompile/keymap/keymap.cpp

```diff
@@ -1,463 +1,861 @@
+        KMX_get_KeyValUnderlying_From_KeyCodeUnderlying(keymap, all_vector[0][i][0], 1);  // shift state: shifted:1
   }
-  // ? CHECK if ran OK, vector size is correct -> return 0/1
+
+  return 0;
 }
 
-v_str_2D create_empty_2D( int dim_rows,int dim_shifts)
-{
-  std::string empty = "--";
-  v_str_1D shifts;
-  v_str_2D all;
+bool InitializeGDK(GdkKeymap** keymap, int argc, gchar* argv[]) {
+  // get keymap of underlying keyboard
 
-  for ( int i=0; i< dim_rows;i++) {
-    for ( int j=0; j< dim_shifts;j++) {
-      shifts.push_back(empty);
-    }
-    all.push_back(shifts);
-    shifts.clear();
+  gdk_init(&argc, &argv);
+  GdkDisplay* display = gdk_display_get_default();
+  if (!display) {
+    printf("ERROR: can't get display\n");
+    return 1;
+  }
+
+  *keymap = gdk_keymap_get_for_display(display);
+  if (!keymap) {
+    printf("ERROR: Can't get keymap\n");
+    gdk_display_close(display);
+    return 2;
+  }
+  return 0;
+}
+
+bool IsKeymanUsedChar(int KV) {
+  //         32            A-Z                      a-z
+  if ((KV == 0x20) || (KV >= 65 && KV <= 90) || (KV >= 97 && KV <= 122))
+    return true;
+  else
+    return false;
+}
+
+std::u16string convert_DeadkeyValues_To_U16str(int in) {
+  if (in == 0)
+    return u"\0";
+
+  if (in < (int)deadkey_min) {                    // no deadkey; no Unicode
+    return std::u16string(1, in);
   }
-  //printf("+++++++ dimensions of Vector after create_empty_2D\t\t %li..%li..%li\n", all.size(), all[0].size(),all[1].size());
-  return all;
+
+  std::string long_name((const char*)gdk_keyval_name(in));  // e.g. "dead_circumflex",  "U+017F",  "t"
+
+  if (long_name.substr(0, 2) == "U+")             // U+... Unicode value
+    return CodePointToU16String(in - 0x1000000);  // GDK's gdk_keymap_translate_keyboard_state() returns (Keyvalue | 0x01000000)
+                                                  // since we never have a carry-over we can just subtract 0x01000000
+
+  KMX_DWORD lname = convertNamesTo_DWORD_Value(long_name);  // 65106 => "dead_circumflex" => 94 => "^"
+
+  if (lname != INVALID_NAME) {
+    return std::u16string(1, lname);
+  } else
+    return u"\0";
 }
 
-int GetKeyvalsFromKeymap(GdkKeymap *keymap, guint keycode, int shift_state_pos) {
-  GdkKeymapKey *maps;
-  guint *keyvals;
+int KMX_get_KeyVal_From_KeyCode(GdkKeymap* keymap, guint keycode, ShiftState ss, int caps) {
+  GdkModifierType consumed;
+  GdkKeymapKey* maps;
+  guint* keyvals;
   gint count;
-  int out;
 
   if (!gdk_keymap_get_entries_for_keycode(keymap, keycode, &maps, &keyvals, &count))
     return 0;
-  //if(!gdk_wayland_keymap_get_entries_for_keycode(keymap, keycode, &maps, &keyvals, &count))
-  //  return 0;
 
-  if (!(shift_state_pos < count))
+  if (!(ensureValidInputForKeyboardTranslation(convert_WinShiftstate_to_LinuxShiftstate(ss), (int)count, (int)keycode))) {
+    g_free(keyvals);
+    g_free(maps);
     return 0;
+  }
 
-  out = keyvals[shift_state_pos];
+  // BASE (shiftstate: 0)
```

Comment:
Hmm, I'm wondering if we have to use `convert_WinShiftstate_to_LinuxShiftstate(ss)` instead of directly `ss`. Otherwise calling `convert_WinShiftstate_to_LinuxShiftstate(ss)` when calling `ensureValidInputForKeyboardTranslation` doesn't make much sense to me.